### PR TITLE
https support

### LINKED ISSUE #369: New feature: HTTPS option in settings
Please can you add option for secure communication in settings? When client communicate over HTTP, then communication is unsecured. When a Apache web server is built in front of TVheadend server as HTTPS proxy, HTTP communication must be redirected to HTTPS port. Unfortunately, the client is able to send the login information before the communication is redirected to the HTTPS protocol. The login information can easily be intercepted during initial communication via the HTTP protocol. If communication could be started directly on the HTTPS protocol, communication would be secured and the login data could not be intercepted. The HTTPS option is great possibilty how to secure TV streaming outside LAN.
How to built HTTPS proxy for TVheadend:
https://tvheadend.org/boards/5/topics/20739
https://tvheadend.org/issues/4372
https://tvheadend.org/boards/5/topics/27593

Complete Apache config:
```
<VirtualHost *:443>
  SSLEngine on
  ServerName tvh.server.net:443
  SSLCertificateFile "c:\Users\Administrator\AppData\Roaming\crt\tvh.server.net-chain.pem"
  SSLCertificateKeyFile "c:\Users\Administrator\AppData\Roaming\crt\tvh.server.net-key.pem"
  CustomLog "${SRVROOT}/logs/tvh-server-net_ssl_request.log" \
          "%t %h %{SSL_PROTOCOL}x %{SSL_CIPHER}x \"%r\" %b"

ProxyRequests Off
ProxyPreserveHost On

ProxyPass /comet/ws ws://192.168.0.50:9981/comet/ws
ProxyPassReverse /comet/ws ws://192.168.0.50:9981/comet/ws

ProxyPass / http://192.168.0.50:9981/
ProxyPassReverse / http://192.168.0.50:9981/

AddOutputFilterByType INFLATE;SUBSTITUTE;DEFLATE audio/x-mpegurl
Substitute "s|http://|https://|i"
</virtualhost> 


Redirect:

<VirtualHost *:80>
    ServerName tvh.server.net
    ErrorLog logs/tvh.server.net-error_log
    CustomLog logs/tvh.server.net-access_log common

    Redirect / https://tvh.server.net/
</VirtualHost>
```

### PR DESCRIPTION
**Pull request by rubdos**
Will eventually resolve #369.

### CODE REVIEW

File: src/tvheadend/HTSPConnection.cpp [hunk 1]
```diff
@@ -144,7 +144,13 @@ std::string HTSPConnection::GetWebURL ( const char *fmt, ... ) const
   if (!auth.empty())
     auth += "@";
 
-  std::string url = StringUtils::Format("http://%s%s:%d", auth.c_str(), settings.GetHostname().c_str(), settings.GetPortHTTP());
+  const char *proto = nullptr;
```
Comment (Jalle19) [inline src/tvheadend/HTSPConnection.cpp]: Could be shortened to `const std::string proto = settings.GetUseHTTPS() ? 'https' : 'http';`

**Comment by rubdos** [comment]
I haven 't tested this yet, will do that later... Just wanted this out somewhere already :)

File: src/tvheadend/Settings.h [hunk 2]
```diff
@@ -106,6 +107,7 @@ namespace tvheadend {
     : m_strHostname(DEFAULT_HOST),
       m_iPortHTSP(DEFAULT_HTTP_PORT),
       m_iPortHTTP(DEFAULT_HTSP_PORT),
+      m_buseHTTPS(DEFAULT_USE_HTTPS),
```
Comment (Jalle19) [inline src/tvheadend/Settings.h]: Wrong member name here, should be `m_bUseHTTPS`

Comment (rubdos) [reply src/tvheadend/Settings.h]: I should really compile this before pushing... :'-)

File: src/tvheadend/Settings.h [hunk 3]
```diff
@@ -38,6 +38,7 @@ namespace tvheadend {
     static const std::string DEFAULT_HOST;
     static const int         DEFAULT_HTTP_PORT;
     static const int         DEFAULT_HTSP_PORT;
+    static const bool         DEFAULT_USE_HTTPS;
```
Comment (ksooo) [inline src/tvheadend/Settings.h]: wrong indentation

Comment (rubdos) [reply src/tvheadend/Settings.h]: Slip of the eye. It compiles now, but I've not been able to test yet. Gotta get something x86 with X11 for that.

**Review by Jalle19 — APPROVED** [review]

**Comment by rubdos** [comment]
Mind that I still didn't have a chance to test this, so please do so before merging :'-)

**Comment by rubdos** [comment]
I tested it working, but the text for the button in the settings is still empty.

**Comment by rubdos** [comment]
> I tested it working, but the text for the button in the settings is still empty.

Could someone give me a hint how I can update the necessary translation files for this?

**Comment by Gee1111** [comment]
this also works for the windows addon ?

**Comment by rubdos** [comment]
> this also works for the windows addon ?

I wouldn't know how that'd be different. But hey, it's Windows, better test it :-)

**Comment by Jalle19** [comment]
> Could someone give me a hint how I can update the necessary translation files for this?

You only need to update en_GB, the rest are handled later on through Transifex. If you're using another language in Kodi you might wanna switch to the en_GB while developing so you can see your changes.

File: pvr.hts/resources/settings.xml [hunk 4]
```diff
@@ -2,6 +2,7 @@
 <settings>
   <category label="30000">
     <setting id="host" type="text" label="30001" default="127.0.0.1" />
+    <setting id="https" type="bool" label="30018" default="false" />
```
Comment (ksooo) [inline pvr.hts/resources/settings.xml]: Please use id 30011 and add a string for it here https://github.com/kodi-pvr/pvr.hts/blob/master/pvr.hts/resources/language/resource.language.en_gb/strings.po#L65

**Review by ksooo — CHANGES_REQUESTED** [review]
please change settings id to first available general settings id (30011) and supply a string for it.

**Comment by ksooo** [comment]
@rubdos have you seen that we requested changes?

File: pvr.hts/resources/settings.xml (hunk 4 above)
Comment (rubdos) [reply pvr.hts/resources/settings.xml]: Done.

**Comment by rubdos** [comment]
Excuse me for the late reply. Have been quite busy.

W.r.t. base branch: are you sure we don't want to backport this into Leia? It's quite an annoying bug imo.

**Comment by jobrien2001** [comment]
@ksooo Any chance to review this? :D

**Comment by rubdos** [comment]
Apparently I never fixed the translation ID. That's done now. I also rebased against Matrix.

**Comment by trimethia** [comment]
@ksooo Any chance to review this?

**Comment by AlwinEsch** [comment]
Related to https://github.com/xbmc/xbmc/pull/17307, not forget a rebase of request to allow further Jenkins use!

```sh
git remote add upstream https://github.com/kodi-pvr/pvr.hts
git fetch upstream
git rebase upstream/Matrix
```

**Comment by rubdos** [comment]
It's only the third time I'm rebasing this now, I would really appreciate to know what still has to be done to get this merged.

**Comment by ksooo** [comment]
Please update the changelog and increase addon minor version and we are ready to go.

**Comment by ksooo** [comment]
> Please update the changelog and increase addon minor version and we are ready to go.

 Well not really. Kodi default language is en_gb, not en_us. So, please add the new settings string to pvr.hts/resources/language/resource.language.en_gb/strings.po and remove it from pvr.hts/resources/language/resource.language.en_us/strings.po as   translations are handled via Transifex, not Github.

**Review by ksooo — CHANGES_REQUESTED** [review]
See my comments

**Comment by rubdos** [comment]
> See my comments

I don't see any comments in the diff linked by "View changes".

I've changed the translation to `en_gb`, added the changelog entry, incremented the minor version. Would you like me to squash everything together?

**Comment by ksooo** [comment]
Yes, please squash to a single commit.

File: pvr.hts/addon.xml.in [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <addon
   id="pvr.hts"
-  version="5.3.4"
+  version="5.3.5"
```
Comment (ksooo) [inline pvr.hts/addon.xml.in]: Please increase the minor version number, not the micro as we are adding a new feature with this PR. => 5.4.0

File: src/tvheadend/HTSPConnection.cpp [hunk 6]
```diff
@@ -149,7 +149,8 @@ std::string HTSPConnection::GetWebURL(const char* fmt, ...) const
   if (!auth.empty())
     auth += "@";
 
-  std::string url = StringUtils::Format("http://%s%s:%d", auth.c_str(),
+  const char *proto = settings.GetUseHTTPS() ? "https" : "http";
```
Comment (ksooo) [inline src/tvheadend/HTSPConnection.cpp]: Please left align the asterisk. Coding guide lines. => 'const char* proto'

**Review by ksooo — CHANGES_REQUESTED** [review]

**Comment by rubdos** [comment]
Comments addressed, squashed!

**Review by ksooo — APPROVED** [review]
Thanks for your contribution.

**Pull request merged**